snap/integrity/integrity.go

```diff
@@ -0,0 +1,284 @@
+	Offset         uint64
+	SourceFilePath string
+}
+
+// FindIntegrityData returns integrity data information given a snap filename.
+// It currently supports only integrity data attached at the end of snap files.
+func FindIntegrityData(snapName string) (*IntegrityData, error) {
+	integrityData := IntegrityData{}
+
+	snapFile, err := os.OpenFile(snapName, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer snapFile.Close()
+
+	snapFileInfo, err := snapFile.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	if !squashfs.FileHasSquashfsHeader(snapName) {
+		return nil, fmt.Errorf("input file does not contain a SquashFS filesystem")
+	}
+
+	// Seek to bytes_used field of SquashFS superblock
+	_, err = snapFile.Seek(squashfsSuperblockBytesUsedOffset, io.SeekStart)
+
+	var squashFSSize uint64
+	if err := binary.Read(snapFile, binary.LittleEndian, &squashFSSize); err != nil {
+		return nil, err
+	}
+
+	logger.Debugf("SquashFS bytes used: %d", squashFSSize)
+
+	// Align squashFSSize to blockSize
+	offset := align(squashFSSize)
+
+	if offset == uint64(snapFileInfo.Size()) {
+		return nil, fmt.Errorf("Integrity data not found for snap %s", snapName)
+	}
+
+	integrityData.SourceFilePath = snapName
+	// TODO check for integrity data in separate file
+
+	_, err = snapFile.Seek(int64(offset), io.SeekStart)
+
+	integrityDataBytes := make([]byte, uint64(HeaderSize))
+	n, err := snapFile.Read(integrityDataBytes)
+	if n < HeaderSize {
+		return &integrityData, fmt.Errorf("failed to read integrity data: integrity data header corrupted?")
+	}
+
+	integrityDataHeader, err := ExtractIntegrityDataHeader(integrityDataBytes)
+	if err != nil {
+		return nil, err
+	}
+	integrityData.Header = integrityDataHeader
+	integrityData.Offset = offset
+
+	return &integrityData, nil
+}
+
+// Validates integrity data against a snap-revision assertion.
+// The entire integrity data including the header get hashed and compared to the hash
+// included in the integrity stanza in a snap-revision assertion.
+func (integrityData IntegrityData) Validate(snapRev asserts.SnapRevision) error {
+	integrityDataHash, err := integrityData.SHA3_384()
+	if err != nil {
+		return err
+	}
+
+	snapRevIntegrityData := snapRev.SnapIntegrity()
+	if snapRevIntegrityData == nil {
+		return fmt.Errorf("Snap revision assertion does not contain an integrity stanza")
+	}
+
+	if integrityDataHash != snapRevIntegrityData.SHA3_384 {
+		return fmt.Errorf("integrity data hash mismatch")
+	}
+	return nil
+}
+
+// Parses a byte array containing integrity data header information into an IntegrityDataHeader
+// structure.
+func ExtractIntegrityDataHeader(integrityDataBytes []byte) (*IntegrityDataHeader, error) {
+	integrityDataHeader := IntegrityDataHeader{}
+
+	err := integrityDataHeader.Decode(integrityDataBytes[:HeaderSize])
+	if err != nil {
+		return nil, err
+	}
+
+	return &integrityDataHeader, nil
+}
+
+// Computes the SHA3-384 digest of the integrity data and encodes it in the hash format used
```

Comment:
same, doc comment need to start with the method name

Reply:
Fixed.